### DISCUSSION TITLE
Time gating in latest scikit-rf version

### DISCUSSION THREAD

**Discussion by brazensky** [scikit-rf/scikit-rf#891] [Q&A] [answered]
Hi all!

I was using time gating for some VNA measurements, and I noticed that the command has changed lately.

In the first attached picture, you see how time gated data look now, with the overshoot at the ends of the band. I plot S11 of a passive device here.
In the second attached picture, you see the same data with the same time gating but in an older version of scikit-rf. Here there is no Gibbs phenomena at the ends of the band.
Window used is ('kaiser', 6) in both cases; I tried other windows as well but the overshoot remains.

Is there any way to fix this in the newest version of scikit-rf?

Thank you in advance :)

![tgnew](https://user-images.githubusercontent.com/130503329/231410495-f07af1ed-0de5-4891-aa81-9f5ee539ecea.png)
![tgold](https://user-images.githubusercontent.com/130503329/231410501-f80885cf-8dcd-40d4-8192-8915414bae6a.png)

**Comment by Vinc0110** [comment]
Hi!
The gating method has indeed changed in version 0.24.1.
By default, it now uses the FFT method for performance reasons. Before, it was using convolution by default. You can change the method back to convolution with the respective argument:
`time_gate(nw, ...., method='convolution')`

Let us know if this works as expected.

Also see docs for more information: https://scikit-rf.readthedocs.io/en/latest/api/generated/skrf.time.time_gate.html#skrf.time.time_gate

**Reply by brazensky** [reply]
Thank you for your reply, Vincent!

![testgate](https://user-images.githubusercontent.com/130503329/232927493-9ed6bfa9-5a94-4d95-897e-0f2140c20bcb.png)

I have made a comparison for my data using the older and newer version of scikit rf. In the attached picture, I include the data before time gating for reference, the gated data using the older version of scikit rf and the gated data using the newer scikit rf version, for both the fft and the convolution method. Even for the convolution method, the two versions produce different results, which are more apparent at the ends of the band. I think the older version (orange curve) still represents the original data more accurately at the band's edges, without any overshoots. Green curve (new version, convolution method) still gives some overshoot, even though not as pronounced as the fft method. All other parameters are the same for the 3 different curves.

**Reply by Vinc0110** [reply, marked as answer]
Thanks for the tests!

The roll off at the band edges is a known issue. It seems like that behavior depends on the actual data. That's why I added two more arguments to `time_gate()` to allow for more flexibility. It would be interesting if you could run your tests a few more times with different settings:
- argument `conv_mode `: It affects the treatment of the band edges in **convolution mode**. Previously, this was set to `'reflect'`, now the default is `'wrap'`. That argument gets passed to `scipy.ndimage.convolve1d()` as 'mode', see https://docs.scipy.org/doc/scipy/reference/generated/scipy.ndimage.convolve1d.html
- argument `fft_window`: It affects the frequency-domain windowing in **FFT mode**. This is different from the time-domain windowing configured by the `window` argument, where you are using the Kaiser window. The default is `fft_window='cosine'`, but you can try different windows or `None`. It gets passed to `scipy.window.get_window()`, see https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.get_window.html

I'm curious to see your results.

**Reply by brazensky** [reply]
Setting `conv_mode ` to `reflect` seems to be almost identical to the previous version, without roll off at the edges. I have many measurements however, so I will test it with many different data to see how it depends on the actual data itself!

Thank you again for the clarification!
